text/0000-portability-lint.md

```diff
@@ -0,0 +1,574 @@
+built-in portabilities, which are fed to the SAT solver when comparing
+portabilities. The end result is that code like the following should pass the lint:
+
+```rust
+#[cfg(unix)]
+fn unix_only() { .. }
+
+#[cfg(linux)]
+fn linux_only() {
+    // permitted since `linux` implies `unix`
+    unix_only()
+}
+```
+
+### Determining the portability of referenced items
+
+**How is the portability of a referenced item determined?** The lint will simply
+resolve an item to its definition, and use the portability of that
+definition. In particular, when invoking trait items, the portability will be
+that of the item's *definition*, rather than that of the `impl`. See
+[Unresolved questions][unresolved] for more on this point; it's something that
+could be revised over time.
+
+#### `match_cfg`
+
+The `match_cfg` macro takes a sequence of `cfg` patterns, followed by `=>` and
+an expression. Its syntax and semantics resembles that of `match`. However,
+there are some special considerations when checking portability:
+
+* When descending into an arm of a `match_cfg`, the arm is checked against
+  portability that includes the pattern for the arm.
+
+* The portability for the `match_cfg` itself is understood as `any(p1, ...,
+  p_n)` where the `match_cfg` patterns are `p1` through `p_n`.
+
+Thus, for example, the following code will pass the lint:
+
+```rust
+#[cfg(windows)]
+fn windows_only() { .. }
+
+#[cfg(unix)]
+fn unix_only() { .. }
+
+#[cfg(any(windows, unix))]
+fn portable() {
+    // the expression here has portability `any(windows, unix)`
+    match_cfg! {
+        windows => {
+            // allowed because we are within a scope with
+            // portability `all(any(windows, unix), windows)`
+            windows_only()
+        }
+        unix => {
+            // allowed because we are within a scope with
+            // portability `all(any(windows, unix), unix)`
+            unix_only()
+        }
+    }
+}
+```
+
+If you have a `match_case` that covers *all* cases (like `windows` and
+`not(windows)`), then it imposes *no* portability constraints on its context.
+
+## The story for `std`
+
+With these basic mechanisms in hand, let's sketch out how we might apply them to
+the standard library to achieve our initial goals. This part of the RFC should
+not be considered normative; it's left to the implementation to make the final
+determination about how to set up the standard library.
+
+### The mainstream platform
+
+The "mainstream platform" will be expressed via a new primitive `cfg` pattern
+called `std`. This is the **default portability of all crates**, unless
+opted-out (see below on "subsetting `std`"). Likewise, most items in `std` will
+initially be exported at `std` portability level. These two facts together mean
+that existing uses of `std` will continue to work without issuing any warnings.
+
+The `std` portability will include several implications, e.g.:
+
+- `std` implies `any(windows, macos, linux)`
```

Comment:
It seems to me like this implication would make *any* crate with `target_os ≠ any("windows", "macos", "linux")` fire the lint on any crate whenever used in code specific to non-currently-tier-1 platform.

For example this code:

```
#[cfg(target_os = "freebsd")]
fn new_vec() -> Vec<u8> { Vec::new() }
```

So the body here is `windows | macos | linux` but the function itself is `target_os = "freebsd"`. Since `target_os = "freebsd"` is not implied by any of the `windows`, `macos` or `linux`, `Vec::new()` is considered to be incompatible and gets linted. Am I right? If so, that sounds like a serious issue.

The way I see it, std portability should be derived from every supported target. Looking at [platform support page](https://forge.rust-lang.org/platform-support.html), there’s a number of targets supported, I’ll pick a few:

* `x86_64-unknown-linux-gnu` = `all(target_arch = "x86", target_os= "linux", target_vendor = "unknown", target_abi = "gnu", target_pointer_width = "64", ...)`,
* `mips64-unknown-linux-gnuabi64` = `all(target_arch = "mips64", target_os = "linux", target_vendor = "unknown", target_abi = "gnuabi64", target_pointer_width = "64", ...)`,
* `thumbv7m-none-eabi` (*std not supported*) = `not(all(target_arch = "...", target_os = "none", ...))`,
* ...

and if *any* of those conditions above hold, then the `std` portability is implied.

Now the question how do you specify it? `#![cfg(...)]` of all these on `src/libstd/lib.rs` makes no sense as there’s also a number of custom targets (e.g. `x86_64-unknown-linux-{arbitrarylibcabi}`) on which libstd runs fine… probably… and wouldn’t anymore.

---

Eh, I wrote this whole comment and only remembered about non-normative nature of this section.